[16.0][IMP] sale_stock_available_to_promise_release_block: improve Unblock Release process

sebalix opened the pull request:
Replace the `Unblock Release` server action by a wizard.

This new wizard allows to give more options to the user regarding the unblocking process, like the scheduled date to set on unblocked moves, and re-assign automatically a stock operation on them (so they could be grouped together in the same transfer).
Also, blocked delivery moves can be planned for unblocking when the current SO is confirmed.

TODO:
- [x] <s>handle release channel to propagate on stock operations:</s> won't do, there won't be direct dependencies of sale against  release channels. Instead a preferred release channel for a given delivery date can be set by sale users thanks to #838 , so the right release channel is selected by the usual mechanism of `stock_release_channel`.

sebalix committed:
sale_stock_available_to_promise_release: rework tests.common.Common class to ease extension

sebalix committed:
sale_stock_available_to_promise_release_block: add Unblock Release wizard

This new wizard allows to give more options to the user regarding the
unblocking process, like the scheduled date to set on unblocked moves,
and re-assign automatically a stock operation on them (so they could be
grouped together in the same transfer).

jbaudoux commented on sale_stock_available_to_promise_release_block/wizards/unblock_release.py:
```diff
@@ -0,0 +1,103 @@
+    _sql_constraints = [
+        (
+            "check_scheduled_date",
+            "CHECK (date_deadline::date >= now()::date)",
```
not timezone friendly...

sebalix replied:
Indeed, I'll change this to a Python constraint

jbaudoux commented on sale_stock_available_to_promise_release_block/wizards/unblock_release.py:
```diff
@@ -0,0 +1,103 @@
+            ("asap", "As soon as possible"),
+        ]
+        if self.env.context.get("from_sale_order_id"):
+            options.append(("contextual", "Contextual"))
```
```suggestion
            options.append(("contextual", "From contextual sales order"))
```

jbaudoux commented on sale_stock_available_to_promise_release_block/wizards/unblock_release.py:
```diff
@@ -0,0 +1,102 @@
+        # Update the scheduled date
+        date_deadline = (
+            fields.Datetime.now() if self.option == "asap" else self.date_deadline
+        )
```
already done in `_compute_date_deadline`
```suggestion
        date_deadline = self.date_deadline
```

jbaudoux reviewed (commented):
looking good so far

sebalix commented:
@jbaudoux reworked to plan the unblocking of undelivered moves at SO confirmation in the case of the "contextual unblocking"

jbaudoux commented on sale_stock_available_to_promise_release_block/wizards/unblock_release.py:
```diff
@@ -0,0 +1,131 @@
+    date_deadline = fields.Datetime(
+        compute="_compute_date_deadline", store=True, readonly=False, required=True
+    )
+    # release_channel_id = fields.Many2one("stock.release.channel", required=True)
```
to remove from here

jbaudoux commented on sale_stock_available_to_promise_release_block/wizards/unblock_release.py:
```diff
@@ -0,0 +1,131 @@
+            group.name += " BACKORDERS"
+            moves.group_id = group
+        # Update the scheduled date and date deadline
+        # TODO: update date_priority to now()?
```
no, don't touch date_priority

jbaudoux commented on sale_stock_available_to_promise_release_block/wizards/unblock_release.py:
```diff
@@ -0,0 +1,131 @@
+        # Re-assign deliveries
+        moves._assign_picking()
+        # Unblock release
+        moves.action_unblock_release()
```
I would move this in the calling method as this does not relate to `_reschedule_moves` but is the next step.

jbaudoux commented on sale_stock_available_to_promise_release_block/wizards/unblock_release.py:
```diff
@@ -0,0 +1,131 @@
+    @api.model
+    def _reschedule_moves(self, moves, date_deadline, from_order=None):
+        """Reschedule the moves based on the deadline."""
+        # Unset current deliveries (keep track of them to delete empty ones at the end)
```
I would here filter out moves that don't have need_release (just to be safe...)

jbaudoux reviewed (approved):
@sebalix Can you squash commits ?

sebalix commented:
@jbaudoux done

gurneyalex committed:
sale_stock_available_to_promise_release_block: onchange speedup

* declare a dependency on the sale.order.line `is_release_blocked` field
to avoid systematic recomputation
* clean the `release_blocked_label` by setting it to `False` instead of an
empty string: this way the field does not show up in the diff when odoo
compares lines in an onchange

jbaudoux commented:
@sebalix precommit is red

jbaudoux commented:
/ocabot merge nobump
